### PR TITLE
Refactor KedroContext with `attrs`

### PR DESCRIPTION
**Pull request by noklam**
Signed-off-by: noklam <nok.lam.chan@quantumblack.com>

## Description

Related to #1465
This change is an intermediate solution for `0.18.x` until we fully make `KedroContext` frozen in `0.19.x`. #1465 was already merged into `develop`, which should not be release until `0.19.0`.

A separate PR will be made for main.
## Development notes

* Refactor `KedroContext` with `attrs`. The interface & logic are unchanged, it should be pure refactoring work.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1566"></a>

### CODE REVIEW

File: tests/framework/context/test_context.py [hunk 1]
```diff
@@ -209,6 +210,15 @@ def test_attributes(self, tmp_path, dummy_context):
         assert isinstance(dummy_context.project_path, Path)
         assert dummy_context.project_path == tmp_path.resolve()
 
+    def test_immutable_instance(self, dummy_context):
```
Comment (AntonyMilneQB) [inline tests/framework/context/test_context.py:213]:
```suggestion
    def test_immutable_attributes(self, dummy_context):
```

**Review by AntonyMilneQB — APPROVED** [review]
Thanks for doing this!

**Comment by sbrugman** [comment]
What's the status of this?

**Comment by noklam** [comment]
@sbrugman sorry for the late response, feel free to tag me if needed. IIRC the implementation works but it's tricky to get the tests to work, due to the mock object missing all the attributes created with attrs or data classes.

**Comment by noklam** [comment]
Closing this as we have a working implementation in `dev` already. It doesn't work on `main` because it use different features of `attrs`. The issue still exists, it's just that in `dev` we somehow avoid dealing with this problem directly.

**Pull request closed without merging**